src/main/java/org/folio/search/service/systemuser/SystemUserService.java

```diff
@@ -0,0 +1,163 @@
+    systemUserParameters.setOkapiToken(backgroundUserApiKey);
+    saveSystemUserParameters(systemUserParameters);
+  }
+
+  private String loginSystemUser(SystemUser params) {
+    var response = authnClient.getApiKey(UserCredentials
+      .of(params.getUsername(), params.getPassword()));
+
+    var headers = response.getHeaders().get(XOkapiHeaders.TOKEN);
+    if (CollectionUtils.isEmpty(headers)) {
+      throw new IllegalStateException(String.format("User [%s] cannot log in", params.getUsername()));
+    } else {
+      return headers.get(0);
+    }
+  }
+
+  private SystemUser buildDefaultSystemUserParameters() {
+    return SystemUser.builder()
+      .id(UUID.randomUUID())
+      .username(folioSystemUserConf.getUsername())
+      .password(folioSystemUserConf.getPassword())
+      .okapiUrl(folioContext.getOkapiUrl())
+      .tenantId(folioContext.getTenantId()).build();
+  }
+
+  private void saveSystemUserParameters(SystemUser systemUserParams) {
+    systemUserRepository.save(systemUserParams);
+  }
+
+  public SystemUser getSystemUserParameters(String tenantId) {
+    return systemUserRepository.getByTenantId(tenantId)
+      .orElseThrow(() -> new IllegalArgumentException("No system user for tenant " + tenantId));
+  }
+
+  private Optional<UsersClient.User> getFolioUser(String username) {
+    var users = usersClient.query("username==" + username);
+    return users.getUsers().stream().findFirst();
+  }
+
+  private String createFolioUser() {
+    final var user = createUserObject();
+    final var id = user.getId();
+    usersClient.saveUser(user);
+    return id;
+  }
+
+  private void saveCredentials(SystemUser parameters) {
+    authnClient.saveCredentials(UserCredentials.of(parameters.getUsername(),
+      parameters.getPassword()));
+
+    log.info("Saved credentials for user: [{}]", parameters.getUsername());
+  }
+
+  private void assignPermissions(String userId) {
+    List<String> perms = getResourceLines(folioSystemUserConf.getPermissionsFilePath());
+
+    if (isEmpty(perms)) {
+      log.warn("No permissions found to assign to user with id: " + userId);
+      return;
+    }
+
+    var permissions = Permissions.of(UUID.randomUUID()
+      .toString(), userId, perms);
+
+    permissionsClient.assignPermissionsToUser(permissions);
+  }
+
+  private void addPermissions(String userId) {
+    var permissions = getResourceLines(folioSystemUserConf.getPermissionsFilePath());
+
+    if (isEmpty(permissions)) {
+      log.warn("No permissions found to assign to user with id: " + userId);
+      return;
+    }
+
+    permissions.forEach(permission -> {
+      var p = new Permission();
+      p.setPermissionName(permission);
+      try {
+        permissionsClient.addPermission(userId, p);
+      } catch (Exception e) {
+        log.info("Error adding permission {} to System User. Permission may be already assigned. Error was {}",
+          permission, e);
+      }
```

Comment:
This is a bit brute force. Maybe retrieving existing perms and compare with the config first.

Reply:
Agree, changed.